File: engine/sonarqube_scanner.py
"""
SonarQube Scanner: Interface with SonarQube Server for SAST analysis.
Supports both Docker-based local scanning and remote API integration.
"""
import os
import json
import time
import logging
import requests
from pathlib import Path
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, asdict

from config import sonar_config

logger = logging.getLogger(__name__)


@dataclass
class SonarQubeIssue:
    """Represents a SonarQube detected issue."""
    issue_key: str
    rule: str
    severity: str  # BLOCKER, CRITICAL, MAJOR, MINOR, INFO
    component: str
    line: Optional[int]
    message: str
    type: str  # BUG, VULNERABILITY, CODE_SMELL, SECURITY_HOTSPOT
    status: str
    resolution: Optional[str]
    effort: str
    debt: str
    tags: List[str]
    cwe: Optional[str] = None
    
    def to_dict(self) -> Dict:
        return asdict(self)


@dataclass
class SonarQubeScanResult:
    """Complete scan result from SonarQube."""
    project_key: str
    analysis_id: str
    status: str
    issues: List[SonarQubeIssue]
    metrics: Dict[str, Any]
    quality_gate: Dict[str, Any]
    scan_duration_ms: int
    timestamp: str
    
    def to_dict(self) -> Dict:
        return {
            "project_key": self.project_key,
            "analysis_id": self.analysis_id,
            "status": self.status,
            "issue_count": len(self.issues),
            "issues": [i.to_dict() for i in self.issues],
            "metrics": self.metrics,
            "quality_gate": self.quality_gate,
            "scan_duration_ms": self.scan_duration_ms,
            "timestamp": self.timestamp
        }


class SonarQubeScanner:
    """Scanner that interfaces with SonarQube for static analysis."""
    
    # CWE to SonarQube rule mapping (common security rules)
    CWE_RULE_MAP = {
        "CWE-79": "java:S5131",  # XSS
        "CWE-89": "java:S3649",  # SQL Injection
        "CWE-22": "java:S2083",  # Path Traversal
        "CWE-78": "java:S2076",  # Command Injection
        "CWE-94": "java:S5135",  # Code Injection
        "CWE-798": "java:S2068",  # Hardcoded credentials
        "CWE-502": "java:S4544",  # Insecure deserialization
        "CWE-20": "java:S3510",   # Input validation
        "CWE-200": "java:S2127",  # Information exposure
    }
    
    # Simulated SonarQube detection results for demonstration
    SIMULATED_RESULTS = {
        "JAVA-SQLI-001": {
            "detected": True,
            "rule": "java:S3649",
            "severity": "BLOCKER",
            "message": "SQL queries should not be vulnerable to injection attacks",
            "line": 15,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a1", "sql"]
        },
        "JAVA-SQLI-002": {
            "detected": True,
            "rule": "java:S3649",
            "severity": "BLOCKER", 
            "message": "SQL queries should not be vulnerable to injection attacks",
            "line": 12,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a1", "sql"]
        },
        "JAVA-XSS-001": {
            "detected": True,
            "rule": "java:S5131",
            "severity": "CRITICAL",
            "message": "Reflected cross-site scripting vulnerability",
            "line": 10,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a7", "xss"]
        },
        "JAVA-SECRETS-001": {
            "detected": True,
            "rule": "java:S2068",
            "severity": "BLOCKER",
            "message": "Credentials should not be hard-coded",
            "line": 5,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a2", "credentials"]
        },
        "JAVA-PATH-001": {
            "detected": True,
            "rule": "java:S2083",
            "severity": "CRITICAL",
            "message": "Path traversal vulnerability detected",
            "line": 8,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a5", "path"]
        },
        "JAVA-CMDI-001": {
            "detected": False,  # SonarQube may miss this pattern
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        },
        "JAVA-SAFE-001": {
            "detected": False,  # False positive - secure code
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        },
        "JAVA-SAFE-002": {
            "detected": False,
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        },
        "PY-SQLI-001": {
            "detected": True,
            "rule": "python:S3649",
            "severity": "BLOCKER",
            "message": "SQL queries should not be vulnerable to injection attacks",
            "line": 8,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a1", "sql"]
        },
        "PY-CMDI-001": {
            "detected": True,
            "rule": "python:S2076",
            "severity": "BLOCKER",
            "message": "OS commands should not be vulnerable to injection attacks",
            "line": 7,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a1", "command"]
        },
        "PY-SECRETS-001": {
            "detected": True,
            "rule": "python:S2068",
            "severity": "BLOCKER",
            "message": "Credentials should not be hard-coded",
            "line": 4,
            "type": "VULNERABILITY",
            "tags": ["cwe", "owasp-a2", "credentials"]
        },
        "PY-SSRF-001": {
            "detected": False,  # May not be detected by SonarQube
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        },
        "PY-SAFE-001": {
            "detected": False,
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        },
        "PY-SAFE-002": {
            "detected": False,
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        },
        "C-BUF-001": {
            "detected": True,
            "rule": "c:S5907",
            "severity": "BLOCKER",
            "message": "Buffer operations should not be vulnerable to buffer overflow",
            "line": 6,
            "type": "VULNERABILITY",
            "tags": ["cwe", "buffer"]
        },
        "C-FMT-001": {
            "detected": True,
            "rule": "c:S5478",
            "severity": "BLOCKER",
            "message": "Format strings should not be vulnerable to injection",
            "line": 5,
            "type": "VULNERABILITY",
            "tags": ["cwe", "format-string"]
        }
    }
    
    def __init__(self, config=None):
        self.config = config or sonar_config
        self.session = requests.Session()
        
    def _make_request(self, endpoint: str, params: Dict = None) -> Optional[Dict]:
        """Make authenticated request to SonarQube API."""
        url = f"{self.config.host_url}/api/{endpoint}"
        try:
            resp = self.session.get(
                url, 
                params=params,
                headers=self.config.auth_header,
                timeout=30
            )
            if resp.status_code == 200:
                return resp.json()
            else:
                logger.warning(f"SonarQube API returned {resp.status_code}: {resp.text}")
                return None
        except Exception as e:
            logger.error(f"Failed to connect to SonarQube: {e}")
            return None
    
    def check_health(self) -> bool:
        """Check if SonarQube server is accessible."""
        result = self._make_request("system/status")
        if result and result.get("status") == "UP":
            logger.info(f"SonarQube {result.get('version')} is running")
            return True
        return False
    
    def get_project_issues(self, project_key: str) -> List[SonarQubeIssue]:
        """Retrieve issues for a project."""
        issues = []
        page = 1
        page_size = 100
        
        while True:
            result = self._make_request("issues/search", {
                "componentKeys": project_key,
                "ps": page_size,
                "p": page,
                "types": "VULNERABILITY,SECURITY_HOTSPOT"
            })
            
            if not result or "issues" not in result:
                break
                
            for item in result["issues"]:
                issue = SonarQubeIssue(
                    issue_key=item.get("key", ""),
                    rule=item.get("rule", ""),
                    severity=item.get("severity", "INFO"),
                    component=item.get("component", ""),
                    line=item.get("line"),
                    message=item.get("message", ""),
                    type=item.get("type", "CODE_SMELL"),
                    status=item.get("status", "OPEN"),
                    resolution=item.get("resolution"),
                    effort=item.get("effort", ""),
                    debt=item.get("debt", ""),
                    tags=item.get("tags", [])
                )
                issues.append(issue)
            
            if len(result["issues"]) < page_size:
                break
            page += 1
            
        return issues
    
    def scan_sample(self, sample_id: str, source_code: str, language: str) -> Dict:
        """Simulate SonarQube scanning a single code sample.
        
        In production, this would:
        1. Write source to file
        2. Trigger sonar-scanner CLI
        3. Wait for analysis completion
        4. Retrieve issues via API
        """
        result = self.SIMULATED_RESULTS.get(sample_id, {
            "detected": False,
            "rule": None,
            "severity": None,
            "message": None,
            "line": None,
            "type": None,
            "tags": []
        })
        
        # Add metadata
        result["sample_id"] = sample_id
        result["scanner"] = "SonarQube"
        result["scan_duration_ms"] = 150  # Simulated
        result["timestamp"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        
        return result
    
    def get_security_metrics(self, project_key: str) -> Dict[str, Any]:
        """Get security-focused metrics for a project."""
        metrics_list = [
            "vulnerabilities", "security_hotspots", "security_rating",
            "security_remediation_effort", "code_smells", "bugs",
            "coverage", "duplicated_lines_density"
        ]
        
        result = self._make_request("measures/component", {
            "component": project_key,
            "metricKeys": ",".join(metrics_list)
        })
        
        if result and "component" in result:
            measures = result["component"].get("measures", [])
            return {m["metric"]: m["value"] for m in measures}
        return {}


def run_sonarqube_analysis(samples: List[Dict]) -> List[Dict]:
    """Run SonarQube analysis on a list of code samples.
    
    Args:
        samples: List of sample dicts with 'sample_id', 'source_code', 'language'
        
    Returns:
        List of analysis results
    """
    scanner = SonarQubeScanner()
    results = []
    
    for sample in samples:
        result = scanner.scan_sample(
            sample["sample_id"],
            sample["source_code"],
            sample["language"]
        )
        results.append(result)
    
    return results


if __name__ == "__main__":
    scanner = SonarQubeScanner()
    # Test with a sample
    test_result = scanner.scan_sample(
        "JAVA-SQLI-001",
        "test code",
        "Java"
    )
    print(json.dumps(test_result, indent=2))

File: backend/sonarqube-scanner.ts
/**
 * SonarQube Scanner: Interface with SonarQube Server for SAST analysis.
 * Supports both Docker-based local scanning and remote API integration.
 */
import axios, { AxiosInstance } from "axios";
import { sonarConfig } from "./config";
import { SonarQubeIssue, SonarQubeScanResult } from "./types";

interface SimulatedResult {
  detected: boolean;
  rule?: string;
  severity?: string;
  message?: string;
  line?: number;
  type?: string;
  tags: string[];
}

const SIMULATED_RESULTS: Record<string, SimulatedResult> = {
  "JAVA-SQLI-001": {
    detected: true,
    rule: "java:S3649",
    severity: "BLOCKER",
    message: "SQL queries should not be vulnerable to injection attacks",
    line: 15,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a1", "sql"],
  },
  "JAVA-SQLI-002": {
    detected: true,
    rule: "java:S3649",
    severity: "BLOCKER",
    message: "SQL queries should not be vulnerable to injection attacks",
    line: 12,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a1", "sql"],
  },
  "JAVA-XSS-001": {
    detected: true,
    rule: "java:S5131",
    severity: "CRITICAL",
    message: "Reflected cross-site scripting vulnerability",
    line: 10,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a7", "xss"],
  },
  "JAVA-SECRETS-001": {
    detected: true,
    rule: "java:S2068",
    severity: "BLOCKER",
    message: "Credentials should not be hard-coded",
    line: 5,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a2", "credentials"],
  },
  "JAVA-PATH-001": {
    detected: true,
    rule: "java:S2083",
    severity: "CRITICAL",
    message: "Path traversal vulnerability detected",
    line: 8,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a5", "path"],
  },
  "JAVA-CMDI-001": {
    detected: false,
    rule: undefined,
    severity: undefined,
    message: undefined,
    line: undefined,
    type: undefined,
    tags: [],
  },
  "JAVA-SAFE-001": {
    detected: false,
    rule: undefined,
    severity: undefined,
    message: undefined,
    line: undefined,
    type: undefined,
    tags: [],
  },
  "JAVA-SAFE-002": {
    detected: false,
    rule: undefined,
    severity: undefined,
    message: undefined,
    line: undefined,
    type: undefined,
    tags: [],
  },
  "PY-SQLI-001": {
    detected: true,
    rule: "python:S3649",
    severity: "BLOCKER",
    message: "SQL queries should not be vulnerable to injection attacks",
    line: 8,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a1", "sql"],
  },
  "PY-CMDI-001": {
    detected: true,
    rule: "python:S2076",
    severity: "BLOCKER",
    message: "OS commands should not be vulnerable to injection attacks",
    line: 7,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a1", "command"],
  },
  "PY-SECRETS-001": {
    detected: true,
    rule: "python:S2068",
    severity: "BLOCKER",
    message: "Credentials should not be hard-coded",
    line: 4,
    type: "VULNERABILITY",
    tags: ["cwe", "owasp-a2", "credentials"],
  },
  "PY-SSRF-001": {
    detected: false,
    rule: undefined,
    severity: undefined,
    message: undefined,
    line: undefined,
    type: undefined,
    tags: [],
  },
  "PY-SAFE-001": {
    detected: false,
    rule: undefined,
    severity: undefined,
    message: undefined,
    line: undefined,
    type: undefined,
    tags: [],
  },
  "PY-SAFE-002": {
    detected: false,
    rule: undefined,
    severity: undefined,
    message: undefined,
    line: undefined,
    type: undefined,
    tags: [],
  },
  "C-BUF-001": {
    detected: true,
    rule: "c:S5907",
    severity: "BLOCKER",
    message: "Buffer operations should not be vulnerable to buffer overflow",
    line: 6,
    type: "VULNERABILITY",
    tags: ["cwe", "buffer"],
  },
  "C-FMT-001": {
    detected: true,
    rule: "c:S5478",
    severity: "BLOCKER",
    message: "Format strings should not be vulnerable to injection",
    line: 5,
    type: "VULNERABILITY",
    tags: ["cwe", "format-string"],
  },
};

export class SonarQubeScanner {
  private config: typeof sonarConfig;
  private client: AxiosInstance;

  constructor(config = sonarConfig) {
    this.config = config;
    this.client = axios.create({
      baseURL: `${config.host_url}/api`,
      timeout: 30000,
      headers: {
        Authorization: `Basic ${Buffer.from(`${config.token}:`).toString("base64")}`,
      },
    });
  }

  private async makeRequest(endpoint: string, params?: Record<string, any>): Promise<any> {
    try {
      const response = await this.client.get(endpoint, { params });
      return response.data;
    } catch (error) {
      console.warn(`SonarQube API request failed: ${endpoint}`, error);
      return null;
    }
  }

  async checkHealth(): Promise<boolean> {
    const result = await this.makeRequest("system/status");
    if (result?.status === "UP") {
      console.log(`SonarQube ${result.version} is running`);
      return true;
    }
    return false;
  }

  async getProjectIssues(projectKey: string): Promise<SonarQubeIssue[]> {
    const issues: SonarQubeIssue[] = [];
    let page = 1;
    const pageSize = 100;

    while (true) {
      const result = await this.makeRequest("issues/search", {
        componentKeys: projectKey,
        ps: pageSize,
        p: page,
        types: "VULNERABILITY,SECURITY_HOTSPOT",
      });

      if (!result?.issues) break;

      for (const item of result.issues) {
        issues.push({
          issue_key: item.key || "",
          rule: item.rule || "",
          severity: item.severity || "INFO",
          component: item.component || "",
          line: item.line,
          message: item.message || "",
          type: item.type || "CODE_SMELL",
          status: item.status || "OPEN",
          resolution: item.resolution,
          effort: item.effort || "",
          debt: item.debt || "",
          tags: item.tags || [],
        });
      }

      if (result.issues.length < pageSize) break;
      page++;
    }

    return issues;
  }

  scanSample(sampleId: string): Record<string, any> {
    const result = SIMULATED_RESULTS[sampleId] || {
      detected: false,
      rule: undefined,
      severity: undefined,
      message: undefined,
      line: undefined,
      type: undefined,
      tags: [],
    };

    return {
      ...result,
      sample_id: sampleId,
      scanner: "SonarQube",
      scan_duration_ms: 150,
      timestamp: new Date().toISOString(),
    };
  }

  async getSecurityMetrics(projectKey: string): Promise<Record<string, any>> {
    const metricsList = [
      "vulnerabilities",
      "security_hotspots",
      "security_rating",
      "security_remediation_effort",
      "code_smells",
      "bugs",
      "coverage",
      "duplicated_lines_density",
    ];

    const result = await this.makeRequest("measures/component", {
      component: projectKey,
      metricKeys: metricsList.join(","),
    });

    if (result?.component?.measures) {
      const measures: Record<string, string> = {};
      for (const m of result.component.measures) {
        measures[m.metric] = m.value;
      }
      return measures;
    }

    return {};
  }
}

export function runSonarQubeAnalysis(samples: Array<{ sample_id: string }>): Array<Record<string, any>> {
  const scanner = new SonarQubeScanner();
  return samples.map((sample) => scanner.scanSample(sample.sample_id));
}
